sbom-utility-scripts/scripts/add-image-reference-script/add_image_reference.py

```diff
@@ -0,0 +1,276 @@
+#!/usr/bin/env python3
+import argparse
+import json
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional
+
+from packageurl import PackageURL
+
+
+@dataclass
+class Image:
+    repository: str
+    name: str
+    digest: str
+    tag: str
+    arch: Optional[str]
+
+    @staticmethod
+    def from_image_index_url_and_digest(
+        image_url_and_tag: str,
+        image_digest: str,
+        arch: Optional[str] = None,
+    ) -> "Image":
+        """
+        Create an instance of the Image class from the image URL and digest.
+
+        Args:
+            image_url_and_tag (str): Image URL in the format 'registry.com/repository/image:tag'.
+            image_digest (str): Manifest digest of the image. (sha256:digest)
+            arch (Optional[str], optional): Image architecture. Defaults to None.
+
+        Returns:
+            Image: An instance of the Image class representing the image.
+        """
+        repository, tag = image_url_and_tag.rsplit(":", 1)
+        _, name = repository.rsplit("/", 1)
+        return Image(
+            repository=repository,
+            name=name,
+            digest=image_digest,
+            tag=tag,
+            arch=arch,
+        )
+
+    @property
+    def digest_algo_cyclonedx(self) -> str:
+        """
+        Get the digest algorithm used for the image in cyclonedx format.
+        The output is in uppercase.
+
+        Returns:
+            str: Algorithm used for the digest.
+        """
+        algo, _ = self.digest.split(":")
+        mapping = {"sha256": "SHA-256", "sha512": "SHA-512"}
+        return mapping.get(algo, algo.upper())
+
+    @property
+    def digest_algo_spdx(self) -> str:
+        """
+        Get the digest algorithm used for the image in SPDX format.
+
+        Returns:
+            str: Algorithm used for the digest in SPDX format.
+        """
+        algo, _ = self.digest.split(":")
+        return algo.upper()
+
+    @property
+    def digest_hex_val(self) -> str:
+        """
+        Get the digest value of the image in hexadecimal format.
+
+        Returns:
+            str: Digest value in hexadecimal format.
+        """
+        _, val = self.digest.split(":")
+        return val
+
+    def purl(self) -> str:
+        """
+        Get the Package URL (PURL) for the image.
+
+        Returns:
+            str: A string representing the PURL for the image.
+        """
+        return PackageURL(
+            type="oci",
+            name=self.name,
+            version=self.digest,
+            qualifiers={"arch": self.arch, "repository_url": self.repository},
```

Comment:
Cross-referencing this with the index_image_sbom script, I think we should drop the `arch` qualifier

The index_image SBOM uses purls like this

```
pkg:oci/name@sha256:{index_digest}?arch=amd64
pkg:oci/name@sha256:{amd64_manifest_digest}

pkg:oci/name@sha256:{index_digest}?arch=s390x
pkg:oci/name@sha256:{s390x_manifest_digest}
```

What we're creating here is
```
pkg:oci/name@sha256:{amd64_manifest_digest}?arch=amd64
```

Which wouldn't match any purl in the index image SBOM

Reply:
That makes sense. Since we don't have an index digest at this point in the pipeline I'll keep the arch-specific manifest and remove the arch qualifier.

Reply:
Fixed